sstables/sstable_set.cc

```diff
@@ -1180,18 +1180,19 @@ compound_sstable_set::bytes_on_disk() const noexcept {
 
 class compound_sstable_set::incremental_selector : public incremental_selector_impl {
     const schema& _schema;
-    const std::vector<lw_shared_ptr<sstable_set>>& _sets;
     std::vector<sstable_set::incremental_selector> _selectors;
 private:
     std::vector<sstable_set::incremental_selector> make_selectors(const std::vector<lw_shared_ptr<sstable_set>>& sets) {
-        return boost::copy_range<std::vector<sstable_set::incremental_selector>>(_sets | boost::adaptors::transformed([] (const auto& set) {
+        // make_incremental_selector clones the underlying sstable_set impl
+        // so we essentially take a snapshot of the compound set contents here.
+        return boost::copy_range<std::vector<sstable_set::incremental_selector>>(sets | boost::adaptors::transformed([] (const auto& set) {
```

Comment:
it doesn't really clone the set, does it? it seems to me make_incremental_selector() only takes a reference of shared ptr, assuming the underlying content is immutable.

Reply:
but I don't think it should. otherwise reads will pay the price.

Reply:
Ah I understand it now. Now we have each set managed by incremental selector taking a copy of _impl, to avoid that problem of range scan referencing outdated set.

Reply:
But once we have compound sets managed under another compound set (thinking forward to tablets, I elabore further on https://github.com/scylladb/scylladb/pull/14043/commits/9977ccb917ff0c34787a95ae7758393f80574a1b#r1313937913), then taking a copy of _impl is not enough, as compound set may still be pointing to stable sstable_set addresses which may change during a range scan.